File: src/jderobot_color_tuner/comm/cameraClient.py
import sys
import os
from .tools import server2int

rosversion = os.environ["ROS_VERSION"]
ser = int(rosversion)

if ( ser == 2):
    import rclpy
    from .ros2.listenerCameraros2 import ListenerCameraros2

if ( ser == 1):
    import rospy
    from .ros.listenerCamera import ListenerCamera


def __getCameraROS2Client(jdrc, prefix):
    '''
    Returns a Camera ROS2 Subscriber

    '''    
    if (ser == 2):
        print("Receiving " + prefix + " Image from ROS2 messages")
        topic = jdrc.getConfig().getProperty(prefix+".Topic")
        client = ListenerCameraros2(topic, jdrc.noderos2)
        return client
    else:
        print(prefix + ": ROS2 msg are diabled for python "+ sys.version_info[0])
        return None

def __getListenerCamera(jdrc, prefix):
    '''
    @return Camera ROS Subscriber

    '''
    if (ser == 1):
        print("Receiving " + prefix + " Image from ROS messages")
        topic = jdrc.getConfig().getProperty(prefix+".Topic")
        client = ListenerCamera(topic)
        return client
    else:
        print(prefix + ": ROS msg are diabled for python "+ sys.version_info[0])
        return None

def __Cameradisabled(jdrc, prefix):
    '''
    Prints a warning that the client is disabled. This function should never be used. Use getCameraClient instead of this

    @param jdrc: Comm Communicator
    @param prefix: name of client in config file

    @return None

    '''
    print( prefix + " Disabled")
    return None

def getCameraClient (jdrc, prefix):
    '''
    Returns a Camera Client.

    @param jdrc: Comm Communicator
    @param prefix: name of client in config file

    @type jdrc: Comm Communicator
    @type name: String

    @return None if Camera is disabled

    '''
    server = jdrc.getConfig().getProperty(prefix+".Server")
    server = server2int(server)

    cons = [__Cameradisabled, __getCameraROS2Client, __getListenerCamera]

    return cons[server](jdrc, prefix)

File: src/jderobot_carviz/comm/cameraClient.cpp
/*
 *  Copyright (C) 1997-2016 JDE Developers Team
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 *  Authors :
 *       Aitor Martinez Fernandez <aitor.martinez.fernandez@gmail.com>
 */
#include "cameraClient.hpp"
#include <iostream>   // std::cout
#include <string>



namespace Comm {

CameraClient*
getCameraClient(Comm::Communicator* jdrc, std::string prefix){
	CameraClient* client = 0;

	int server;
	server = jdrc->getConfig().asInt(prefix+".Server");
	//server = std::stoi(server_name);

	switch (server){
		case 0:
		{
			std::cout << "Camera disabled" << std::endl;
			break;
		}
		case 1:
		{
            #ifdef ROS_H
				std::cout << "Receiving Image from ROS messages" << std::endl;
				std::string nodeName;
				nodeName =  jdrc->getConfig().asStringWithDefault(prefix+".Name", "LaserNode");
				std::string topic;
				topic = jdrc->getConfig().asStringWithDefault(prefix+".Topic", "");
				ListenerCamera* lc;
				lc = new ListenerCamera(0, nullptr, nodeName, topic);
				lc->start();
				client = (Comm::CameraClient*) lc;
            #else
                throw "ERROR: ROS is not available";
            #endif

		 	break;
		}
		
		case 2:
		{
            #ifdef ROS2_H
                std::cout << "Receiving ImageData from ROS messages" << std::endl;

            #else
                throw "ERROR: ROS2 is not available";
            #endif

		 	break;
		}

		default:
		{
			std::cerr << "Wrong " + prefix+".Server property" << std::endl;
			break;
		}

	}

	return client;


}

}//NS
